src/Compilers/CSharp/Test/Symbol/Compilation/CSharpCompilationOptionsTests.cs

```diff
@@ -434,5 +436,50 @@ public void WithCryptoPublicKey()
             Assert.Same(options, options.WithCryptoPublicKey(default(ImmutableArray<byte>)));
             Assert.Same(options, options.WithCryptoPublicKey(ImmutableArray<byte>.Empty));
         }
+
+        [Fact]
+        [WorkItem(22579, "https://github.com/dotnet/roslyn/issues/22579")]
+        public void UncessaryUsingReportedAsErrorWithRuleset()
+        {
+            var source = @"using System;";
+            var ruleset = new Dictionary<string, ReportDiagnostic>() { { "CS8019", ReportDiagnostic.Error } };
+            var compilation = CreateStandardCompilation(source, options: TestOptions.DebugDll.WithSpecificDiagnosticOptions(ruleset));
+
+            using (MemoryStream peStream = new MemoryStream())
+            {
+                var result = compilation.Emit(
+                    peStream: peStream,
+                    pdbStream: null,
+                    metadataPEStream: null,
+                    xmlDocumentationStream: null,
+                    cancellationToken: default,
+                    win32Resources: null,
+                    manifestResources: null,
+                    options: null,
+                    debugEntryPoint: null,
+                    sourceLinkStream: null,
+                    embeddedTexts: null,
+                    testData: null);
+
+                var diagnostics = result.Diagnostics;
+                Assert.False(result.Success);
+
+                var tree = compilation.SyntaxTrees.Single();
+                var model = compilation.GetSemanticModel(tree);
+                var diagnostics2 = model.GetDiagnostics();
+
+                var diagnostics3 = compilation.GetDiagnostics();
+
+                foreach (var diag in new[] { diagnostics, diagnostics2, diagnostics3})
+                {
+                    Assert.Equal(DiagnosticSeverity.Error, diag.Single().Severity);
+                    diag.Verify(
+                        // (1,1): error CS8019: Unnecessary using directive.
+                        // using System;
+                        Diagnostic(ErrorCode.HDN_UnusedUsingDirective, "using System;").WithLocation(1, 1)
+                    );
+                }
```

Comment:
Minor point: Perhaps create a local function and call `CheckDiagnostics(result.Diagnostics); CheckDiagnostics(model.GetDiagnostics()); CheckDiagnostics(compilation.GetDiagnostics());` rather than using `foreach`.